src/community/security/oauth2/src/main/java/org/geoserver/security/oauth2/GeoServerOAuthAuthenticationFilter.java

```diff
@@ -0,0 +1,312 @@
+     */
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+
+        String cacheKey = authenticateFromCache(this, (HttpServletRequest) request);
+
+        if (SecurityContextHolder.getContext().getAuthentication() == null) {
+
+            doAuthenticate((HttpServletRequest) request, (HttpServletResponse) response);
+
+            Authentication postAuthentication = SecurityContextHolder.getContext()
+                    .getAuthentication();
+            if (postAuthentication != null && cacheKey != null) {
+                if (cacheAuthentication(postAuthentication, (HttpServletRequest) request)) {
+                    getSecurityManager().getAuthenticationCache().put(getName(), cacheKey,
+                            postAuthentication);
+                }
+            }
+        }
+
+        chain.doFilter(request, response);
+    }
+
+    @Override
+    public void logout(HttpServletRequest request, HttpServletResponse response,
+            Authentication authentication) {
+
+        request.setAttribute(GeoServerLogoutFilter.LOGOUT_REDIRECT_ATTR,
+                filterConfig.getLogoutUri());
+
+        OAuth2AccessToken token = restTemplate.getOAuth2ClientContext().getAccessToken();
+        if (token != null && token.getTokenType().equals(OAuth2AccessToken.BEARER_TYPE)) {
+
+            restTemplate.getOAuth2ClientContext().setAccessToken(null);
+
+            final AccessTokenRequest accessTokenRequest = restTemplate.getOAuth2ClientContext()
+                    .getAccessTokenRequest();
+            if (accessTokenRequest != null && accessTokenRequest.getStateKey() != null) {
+                restTemplate.getOAuth2ClientContext()
+                        .removePreservedState(accessTokenRequest.getStateKey());
+            }
+        }
+
+        request.getSession().invalidate();
+        response.setStatus(HttpServletResponse.SC_NO_CONTENT);
+        Cookie[] allCookies = request.getCookies();
+
+        for (int i = 0; i < allCookies.length; i++) {
+            String name = allCookies[i].getName();
+            if (name.equalsIgnoreCase("JSESSIONID")) {
+                Cookie cookieToDelete = allCookies[i];
+                cookieToDelete.setMaxAge(-1);
+                cookieToDelete.setPath("/");
+                cookieToDelete.setComment("EXPIRING COOKIE at " + System.currentTimeMillis());
+                response.addCookie(cookieToDelete);
+            }
+        }
+        SecurityContextHolder.getContext().setAuthentication(null);
+    }
+
+    @Override
+    protected void doAuthenticate(HttpServletRequest request, HttpServletResponse response) {
+
+        String principal = null;
+        try {
+            principal = getPreAuthenticatedPrincipal(request, response);
+        } catch (IOException e1) {
+            // TODO Auto-generated catch block
```

Comment:
Do something here, log the exception, rethrow it, but printing it to the std out/err is not a valid option.